src/main/java/io/temporal/samples/encryptedpayloads/CryptDataConverter.java

```diff
@@ -0,0 +1,193 @@
+/*
+ *  Copyright (c) 2020 Temporal Technologies, Inc. All Rights Reserved
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.samples.encryptedpayloads;
+
+import com.google.common.base.Defaults;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.DataConverterException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import java.util.Optional;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+class EncodingKeys {
+  static final String METADATA_ENCODING_KEY = "encoding";
+  static final String METADATA_WRAPPED_ENCODING_KEY = "wrapped-encoding";
+
+  static final String METADATA_ENCODING_RAW_NAME = "binary/plain";
+  static final ByteString METADATA_ENCODING_RAW = ByteString.copyFrom(METADATA_ENCODING_RAW_NAME,
+      StandardCharsets.UTF_8);
+}
+
+public class CryptDataConverter implements DataConverter {
+  private static final String CIPHER = "AES/GCM/NoPadding";
+  private static final int GCM_NONCE_LENGTH_BYTE = 12;
+  private static final int GCM_TAG_LENGTH_BIT = 128;
+  private static final Charset UTF_8 = StandardCharsets.UTF_8;
+
+  private DataConverter converter;
+
+  public CryptDataConverter(DataConverter converter) {
+    this.converter = converter;
+  }
+
+  private SecretKey getKey() {
+    // Key can be fetched from KMS or other secure storage.
+    return new SecretKeySpec("test-key-test-key-test-key-test!".getBytes(UTF_8), "AES");
+  }
+
+  private static byte[] getNonce(int size) {
+    byte[] nonce = new byte[size];
+    new SecureRandom().nextBytes(nonce);
+    return nonce;
+  }
+
+  private byte[] encrypt(byte[] plainData) throws Exception {
+    SecretKey key = this.getKey();
+    byte[] nonce = getNonce(GCM_NONCE_LENGTH_BYTE);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
```

Comment:
Is this a cheap enough operation to execute per payload?

Reply:
It's required to be per payload due to the `nonce`. It must vary per encryption run.